daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ExplicitDisclosureTest.scala

```diff
@@ -0,0 +1,736 @@
+              (TestMod:Key { label = "test-key", maintainers = (TestMod:listOf @Party (TestMod:House {owner} this)) })
+              (\(key: TestMod:Key) -> (TestMod:Key {maintainers} key));
+         };
+
+         val create: Party -> Update (ContractId TestMod:House) =
+           \(party: Party) ->
+             create @TestMod:House TestMod:House { owner = party };
+
+         val exercise: ContractId TestMod:House -> Update Unit =
+           \(contractId: ContractId TestMod:House) ->
+             exercise @TestMod:House Destroy contractId ();
+
+         val fetch_by_id: ContractId TestMod:House -> Update TestMod:House =
+           \(contractId: ContractId TestMod:House) ->
+             fetch_template @TestMod:House contractId;
+
+         val fetch_by_key: Text -> Option Party -> Update <contract: TestMod:House, contractId: ContractId TestMod:House> =
+           \(label: Text) (maintainer: Option Party) ->
+             let key: TestMod:Key = TestMod:Key { label = label, maintainers = (TestMod:optToList @Party maintainer) }
+             in fetch_by_key @TestMod:House key;
+
+         val lookup_by_key: Text -> Option Party -> Update (Option(ContractId TestMod:House)) =
+           \(label: Text) (maintainer: Option Party) ->
+             let key: TestMod:Key = TestMod:Key { label = label, maintainers = (TestMod:optToList @Party maintainer) }
+             in lookup_by_key @TestMod:House key;
+
+         val listOf: forall(t:*). t -> List t =
+           /\(t:*). \(x: t) ->
+             Cons @t [x] (Nil @t);
+
+         val optToList: forall(t:*). Option t -> List t  =
+           /\(t:*). \(opt: Option t) ->
+             case opt of
+                 None -> Nil @t
+               | Some x -> Cons @t [x] (Nil @t);
+       }
+       """
+  )
+  val ledgerParty: IdString.Party = Ref.Party.assertFromString("ledgerParty")
+  val disclosureParty: IdString.Party = Ref.Party.assertFromString("disclosureParty")
+  val contractId: ContractId = Value.ContractId.V1(crypto.Hash.hashPrivateKey("test-contract-id"))
+  val templateId: Ref.Identifier = Ref.Identifier.assertFromString("-pkgId-:TestMod:House")
+  val templateType: Ref.TypeConName = Ref.TypeConName.assertFromString("-pkgId-:TestMod:House")
+  val keyType: Ref.TypeConName = Ref.TypeConName.assertFromString("-pkgId-:TestMod:Key")
+  val ledgerContractKey: GlobalKey = buildContractKey(ledgerParty)
+  val ledgerContractSKey: SValue = buildContractSKey(ledgerParty)
+  val disclosureContractKey: GlobalKey = buildContractKey(disclosureParty)
+  val disclosureContractSKey: SValue = buildContractSKey(disclosureParty)
+  val ledgerContract: Value.VersionedContractInstance = buildContract(ledgerParty)
+  val disclosedContracts =
+    TableDrivenPropertyChecks.Table(
+      ("disclosedContract", "caseLabel"),
+      (buildDisclosedContract(disclosureParty, withHash = false), "disclosedContractNoHash"),
+      (buildDisclosedContract(disclosureParty), "disclosedContractWithHash"),
```

Comment:
This seems invalid: If the template has a key, we should mandate the key not silently accept that people don’t pass the hash. I think we should:
1. Add a test that this gets rejected if you don’t pass a hash and the template defines a key.
2. Change these tests to use different templates for with and without key.

Reply:
Could you help clarify this point please? 

Are we suggesting that when the `DisclosedContract` instance has no key hash, but the template has a key defined on it, then this is a scenario where the evaluation/execution of `SBUFetchKey` and `SBULookupKey` should error? I suppose this is a check that should be performed as part of building the disclosure table (i.e. in `Speedy#buildDiscTable`?

Reply:
Current code changes should now address this issue.